Feat/toast: Toast 컴포넌트 구현

iyu88 opened the pull request:
## 🤠 개요

- #24
- Toast 컴포넌트를 구현했습니다. 
- Toast 을 비롯한 컴포넌트를 띄우기 위한 Portal 컴포넌트를 React Portal 을 사용하여 구현했습니다. 

## 💫 설명

- components/common 아래에 Portal 컴포넌트를 구현했습니다. 
  - Portal로 삼고 싶은 DOM의 id 값을 props로 전달받습니다. 
  - id 를 전달하지 않으면 _portal-root_ 라는 id 를 가진 DOM 을 찾습니다. 

- Toast 컴포넌트를 구현했습니다. 
  - 관련한 Type과 상수는 constants/toast.ts에 정의했습니다.

  - Props는 다음과 같습니다.

    1) kind : Toast의 종류를 설정합니다. 
        - 5종류입니다. info / success / warning / error / default 
        - title, color, mainIcon 에 영향을 줍니다. 
          1) title : kind의 첫글자를 대문자로 바꿔서 표시합니다. default는 _Alarm_ 입니다.
          2) color : color.ts에 kind에 따른 컬러 코드를 추가했습니다. default는 _검정색_ 입니다.
          3) mainIcon : react-icon에서 적절한 아이콘을 선정해보았습니다. 


    2) title : 사용자가 입력한 Toast Message의 큰 제목입니다. 
        - 사용자가 title 을 입력하지 않으면 kind가 표시됩니다. 
        - kind 도 입력하지 않으면 _Alarm_이 표시됩니다.

    3) message (필수) : 실제로 표시되는 메세지입니다. 

    4) vertical (필수) : 수직 방향의 위치를 결정합니다.
        - 2종류입니다. top / bottom
        - vertical 값을 bounce Keyframe 에 props 로 전달하여 애니메이션이 달라집니다. 
           1) top : 메세지가 위에서 아래로 튕깁니다. 
           2) bottom : 메세지가 아래에서 위로 튕깁니다.  

    5) horizontal (필수) : 수평 방향의 위치를 결정합니다. 
        - 3종류입니다. left / center / right

- title 과 message의 역할에 대한 코멘트를 받습니다. 현재 생각한 예시는 다음과 같습니다. 
  - title : Review 요청이 도착했어요.
  - message : Add Review 버튼을 눌러서 리뷰를 시작하세요.

- 현재 템플릿에서 TempButton 으로 명명되어 있는 버튼은 Toast 메세지를 지우는 이벤트를 등록할 예정입니다. 
  - Button 컴포넌트가 완성되면 갈아끼려고 계획했습니다. 
  - 현재는 Alert 창이 뜨는 이벤트와 hover 스타일만 추가해두었습니다.  

## 📷 스크린샷 (Optional)

<details> <img src="https://user-images.githubusercontent.com/31645195/221430860-26dc3fc5-5336-4d0a-8d9d-f55c20b7787a.png" />
<summary>
<h4> Top Left Toast </h4>
</summary>
</details>


<details> <img src="https://user-images.githubusercontent.com/31645195/221430866-bf297a06-45f9-4819-921c-4656b8359b07.png" />
<summary>
<h4> Top Center Toast </h4>
</summary>
</details>

<details> <img src="https://user-images.githubusercontent.com/31645195/221430867-24dbe843-3cb4-41b8-a530-2c477a16fd28.png" />
<summary>
<h4> Top Right Toast  </h4>
</summary>
</details>


<details> <img src="https://user-images.githubusercontent.com/31645195/221430871-173ac8c9-f943-4b27-82ae-477d40d28fbb.png" />
<summary>
<h4> Bottom Left Toast   </h4>
</summary>
</details>


<details> <img src="https://user-images.githubusercontent.com/31645195/221430873-bb3cccf4-2ab2-49a9-b8fd-3f7295553f93.png" />
<summary>
<h4> Bottom Center Toast    </h4>
</summary>
</details>

<details> <img src="https://user-images.githubusercontent.com/31645195/221430875-a8a602ad-63a8-43d4-82a8-9c622ecd3e44.png" />
<summary>
<h4> Bottom Right Toast    </h4>
</summary>
</details>

<details> <img src="https://user-images.githubusercontent.com/31645195/221430878-497f88bb-9f8c-4a7c-9d10-ffb89d92bb6e.png" />
<summary>
<h4> With Title Toast    </h4>
</summary>
</details>

<details> <img src="https://user-images.githubusercontent.com/31645195/221430885-191b09da-b814-41b6-9e9e-2d80dabdd97f.png" />
<summary>
<h4> Long Message Toast  </h4>
</summary>
</details>


<details> <img src="https://user-images.githubusercontent.com/31645195/221430889-b4bdd5a5-d96f-4291-b18e-3020f87ddb4e.png" />
<summary>
<h4> Info Toast  </h4>
</summary>
</details>


<details> <img src="https://user-images.githubusercontent.com/31645195/221430891-8e57132a-3d92-427d-922c-1edf83c0e4fa.png" />
<summary>
<h4> Success Toast  </h4>
</summary>
</details>

<details> <img src="https://user-images.githubusercontent.com/31645195/221430895-1ed3ae88-dcfa-4fe5-b6c1-bcf768555ca0.png" />
<summary>
<h4> Warning Toast   </h4>
</summary>
</details>


<details> <img src="https://user-images.githubusercontent.com/31645195/221430897-a3f7d94e-7185-4f47-9faf-b3a35feaca52.png" />
<summary>
<h4> Error Toast   </h4>
</summary>
</details>

iyu88 committed:
feat: #24 Toast 를 위한 Portal 생성

iyu88 committed:
feat: #24 Toast 기본 레이아웃 및 애니메이션 구현

- emotion 에서 객체로 스타일 지정
https://www.daleseo.com/emotion/

- keyframe 에 props 전달하기
https://stackoverflow.com/questions/50802681/how-to-pass-props-to-keyframes-in-styled-component-with-react

iyu88 committed:
feat: #24 Toast 역할에 따른 컬러 코드 추가

기존에 없던 info / success / warning 컬러 코드를 추가합니다.

iyu88 committed:
feat: #24 컬러 코드 적용

아이콘에 컬러 코드를 적용합니다.
primary 유형을 삭제하고 default 아이콘을 Notification 아이콘으로 설정합니다.
이 경우 색상은 별도로 설정하지 않습니다.

iyu88 committed:
feat: #24 배경이 있는 아이콘으로 변경

iyu88 committed:
feat: Portal id 를 props 로 전달

기존 Portal 에서 고정 id 값으로 DOM 요소를 불러왔던 점을 확장성을 고려하여 props 로 받을 수 있게 했습니다.
id 값을 전달하지 않으면 기존처럼 portal-root 를 찾아서 children 을 렌더링합니다.

iyu88 committed:
feat: Toast Props 처리 - title, message

title 과 message props 를 받아서 렌더링합니다.
type 에 따른 아이콘에 전달하는 속성을 별도의 객체로 iconStyle 객체로 분리합니다.
최대 너비를 제한하기 위해서 TypographyGroup 스타일 컴포넌트를 생성합니다.

iyu88 committed:
refactor: #24 Toast 컴포넌트 Type 및 상수 분리

iyu88 committed:
refactor: Toast 타입 수정

객체로 된 CSS 속성을 표현하기 위해서 Interpolation 을 import 합니다.
enum 으로 선언했던 타입을 Type 으로 선언합니다.
아이콘 사이즈를 상수로 표현하기 위한 값을 추가합니다.

iyu88 committed:
refactor: Type 수정 반영 및 CSS 적용

변경한 Type 을 Toast 컴포넌트에 반영했습니다.
CSS 속성을 최상위의 Flexbox 에 적용하도록 하여 불필요한 div 태그를 삭제합니다.

iyu88 committed:
refactor: state 삭제 및 switch-case 함수 생성

기존에 state 를 사용하여 렌더링될 아이콘을 결정했던 것을 별도로 분리한 switch-case 함수로 설정합니다.

iyu88 committed:
feat: Toast 스토리 정의

iyu88 committed:
fix: tsconfig.path 의 Alias 반영

import 경로가 비정상적으로 되어 있는 문제를 해결합니다.

iyu88 committed:
fix: tsconfig.path 의 Alias 반영

dohun31 commented on src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,133 @@
+  str[0].toUpperCase() + str.substring(1);
+
+const getMainIcon = (
+  kind: ToastProps['kind'],
```
컴포넌트를 반환하는거면 그냥 const ToastIcon으로 하고 props로 kind넘겨서 `<ToastIcon kind="뭐뭐" size="" 다른 props>` 이렇게 컴포너트로 사용하면 어떤가요?

iyu88 replied:
Toast/index.tsx 내부에 ToastIcon을 분리했습니다. 
파일을 새로 만들까하다가 타입 관련 import 를 여기저기서 하는 것보다 
index.tsx 에 이미 있으니까 한 파일에서 처리하는게 낫겠다고 생각했어요.

x 아이콘도 넣을까 하다가 추가적인 kind 가 필요할 것 같기도 하고 
어차피 x 아이콘은 고정적으로 한 군데에서 사용하니까 추가하지 않았습니다.

dohun31 commented on src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,133 @@
+      {mainIcon}
+      <TypographyGroup>
+        <Typography variant="subtitle2">
+          {title ?? (kind ? startWithCapitalLetter(kind) : 'Alarm')}
```
레이아웃을 위한 컴포넌트면 Container나 FlexBox 사용하는건 어떠나요?

iyu88 replied:
Container 는 width 가 빵빵해서 그런지 
브라우저 너비에 따라서 좌측 아이콘이 찌그러지는 현상이 있어서 
Flexbox 를 적용했습니다.

dohun31 reviewed (commented):
타이틀이랑 dest 간격 조금만 띄워주세요..ㅎㅎㅎ

prayinforrain commented on src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,133 @@
+        <Typography variant="desc">{message}</Typography>
+      </TypographyGroup>
+      <TempButton onClick={() => alert('Clicked!')}>
+        <MdOutlineCancel size={CLOSE_ICON_SIZE} color={mainColor} />
```
이것도 rem으로 안바꿔도 될까요? 궁금

iyu88 replied:
toast.ts 에서 크기를 export 할 때 pixelToRem 함수를 사용해서 
변환된 값을 내보내도록 변경했습니다.

iyu88 committed:
refactor: ToastIcon 분리 및 Flexbox 레이아웃 적용

ToastIcon을 별도로 분리하고 아이콘에 대한 정보를 받아 SVG 를 반환합니다.
Close 아이콘에 대해서는 추가적인 kind 가 필요할 것 같아 당장은 추가하지 않았습니다.
기존에 TypographyGroup 으로 되어있던 스타일 컴포넌트를 Flexbox 로 변경합니다.

iyu88 committed:
refactor: Icon 크기에 pixelToRem 함수 적용

toast.ts 에서 아이콘 크기를 내보낼 때 pixelToRem 를 사용합니다.
ToastIcon 의 size props 타입을 string 으로 변경합니다.

prayinforrain commented:
그때는 좀 띄엄띄엄 봤는데 지금보니 Portal이라는게 있네요. 처음 보는 개념이에요..  
그런데 구현된 Portal을 다른 곳에서 쓰는것처럼 보이지 않는데 어떤 식으로 활용되는지 설명을 부탁드려도 될까요? 궁금해용

iyu88 commented:
Storybook으로 작업하다보니까 실제 코드에 적용된 부분이 없긴 했네요.
React의 Portal 기능까지 라이브러리에서 제공하고자 하면 지금 있는 Portal 컴포넌트를 그대로 제공해도 괜찮을 것 같긴합니다. 

Portal은 React Root가 아닌 사용자가 지정한 DOM 요소를 부모로 하는 포탈을 생성해서 children 요소를 위치시키는 방식입니다. 
실제 활용처는 Modal이나 지금의 Toast 등 부모 요소 스타일에 영향을 받지 않은 채로 표시하고 싶을 때 사용하면 될 것 같습니다. 
위에서 언급한 두 컴포넌트처럼 화면에 표시되는 위치가 부모 요소에 종속된다기보다 때때로 자유로운 설정 필요할 때 말이죠.

prayinforrain commented:
그럼 대충
```
<Portal>
  <Toast/>
</Portal>
```
이런 식으로 사용하는 것으로 이해하면 될까요??
화면을 덮는 fixed div 위에 올리는거랑 어떤 차이가 있는지 잘 모르겠어요

코드에 문제가 있는건 아니고 그냥 제 궁금증이니 approve 미리 눌러둘게요

se030 commented:
여기에도 Button 없어서 적용 못한 사람이 있네요 ... 하루만 휴가 가지고 .. 그 뒤에 알차게 반성하도록 할게요 ....🤪

se030 commented on src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,141 @@
+  color: string;
+}
+
+const ToastIcon = ({ kind, size, color }: ToastIconProps) => {
```
사소하지만 얘는 Toast 폴더 하위에 Toast/ToastIcon.tsx 로 분리해보면 어떨까요? 그게 모듈 이해하기에 더 좋을 것 같아요!

iyu88 replied:
분리했습니다!
고민했던 부분인데 역시 분리하는게 나은가요?
Toast 내부에서만 사용하는 녀석인데 파일로 또 빼기가 애매했어요.
그래도 폴더로 묶여 있으니 괜찮을 것 같네요~

se030 commented on src/components/@common/Portal/index.tsx:
```diff
@@ -0,0 +1,25 @@
+
+  useEffect(() => {
+    setMounted(true);
+    const portalRoot = document.getElementById(id ?? 'portal-root');
```
CdsProvider 컴포넌트에 포털 root를 넣어서 사용하면 어때요?
cds 쓰려면 항상 써야하는거니까 고유한 id(`cds-portal-root` 같은)를 하나 만들어서 cds의 토스트나 모달같은 컴포넌트들은 걔를 쓰게하면 root가 없는 예외도 어느정도는 커버될 것 같아요!

```tsx
const portalRoot = document.getElementById(id ?? PORTAL_ROOT_ID);
```


```tsx
const CdsProvider = ({ children }: CdsProviderProps) => {
  return (
    <>
      <Global />
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
      <div id="cds-portal-root"></div>
    </>
  );
};
```

iyu88 replied:
Provider 내부에 Portal Root 를 위치시키는거 좋은 아이디어인 것 같아요~

한편 ThemeProvider 바깥에 Portal Root 를 설정하게 되면 
Portal 에 위치하는 컴포넌트들이 Theme 영향을 못받을 것 같아서 

다음과 같이 수정해보았어요.

``` jsx
return (
    <>
      <Global />
      <ThemeProvider theme={theme}>
        {children}
        <div id="cds-portal-root"></div>
      </ThemeProvider>
    </>
  );
```

se030 commented on src/components/Toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,107 @@
+
+const Template: ComponentStory<typeof Toast> = (args) => <Toast {...args} />;
+
+export const TopLeftToast = Template.bind({});
```
그리고 스토리명에서는 컴포넌트 이름 다 제거해주세요 ㅎㅎ #21 에 관련 내용 있는데 회의때 얘기한다는게 깜빡했어요 😅

iyu88 replied:
삭제했습니다! 
훨씬 깔끔해진 것 같아요.
이거는 가이드에 나와있는 내용인가요?
아니면 스스로 생각하시기에 더 좋아보인건가요??

se030 reviewed (changes requested):
수고하셨습니다! 디자인이 너무 예뻐요 👍🏻

수정 몇개 제안했는데 읽어보고 코멘트 주세요.
그리고 혹시 로컬 스토리북 안되는거 고치셨나요? 저만 그런건지 모르겠는데 스토리북에서 토스트가 모바일뷰로 봐야 보이네요 (??)

iyu88 committed:
refactor: ToastIcon 컴포넌트 분리

코드 파악을 손쉽게 하기 위해 ToastIcon 관련 코드를 분리합니다.

iyu88 committed:
refactor: Toast 스토리 수정

스토리 내용만을 파악하기 위해서 스토리에서 컴포넌트 이름을 제거합니다.

iyu88 committed:
refactor: Portal Root 를 CdsProvider 에 포함
